### ISSUE TITLE
formulas with dice have side effects

### ISSUE THREAD

**Issue by jostephd** [wesnoth/wesnoth#3922]
@CelticMinstrel [points out](https://github.com/wesnoth/wesnoth/pull/3800#issuecomment-450673922) that [WFL formulas that uses the dice operator](https://github.com/wesnoth/wesnoth/pull/3800#issuecomment-450349963) "[violates] the general rule that formulas cannot cause side-effects".

**Comment by CelticMinstrel** [comment]
The side-effect is a change to the state of the random number generator, so fixing this would require either removing the dice operator or somehow making a stateless random number generator. @jyrkive has pointed out that a hash function could be useful for the latter.

**Comment by gfgtdf** [comment]
I don't think this side effect is a real problem just something you have to be aware of. Also note that this onyl applies to filters that run in a synced context since random results from a unsynced context come from a differnt generator that is seperate from the gamestate.


Another things that could happen in theory is that if during the current synced action no random seed was dameded yet then the dice operator will demand a new andom seed form the server. And if the player has insane lag the user might click around the map during while waiting for the server-generated seed which then migth trigger umc defined select events that change the gamestate. However this seems too unlikeley to me to be an actual issue.

**Comment by GregoryLundberg** [comment]
What is too unlikely?

That lag can be bad? Seems pretty common to me.

Or that users click around randomly when lagged? Virtually guaranteed.

Or that you need a seed? Must happen or we wouldn't provide for it.

I guess you mean that it seems unlikely you'll have a map with clickables which change game state at the same time you find you don't have a random seed and need one, and lag hits while there's a user with a mouse around?

There's gotta be a corollary to Murphy's Law about this happening, probably when it can do the most damage, or cost the most to repaid.

**Comment by CelticMinstrel** [comment]
A side-effect is a side-effect, regardless of whether it's the synced or unsynced random number generator.

**Comment by gfgtdf** [comment]
well yes but in the unsnced case there is no reason to see problem at all.

> I guess you mean that it seems unlikely you'll have a map with clickables which change game state

This one, umc code is not supposed to change the gamestate from unsnced event (like select). So they should not be surprised about issues arising from it.

**Comment by CelticMinstrel** [comment]
There is still a problem in the unsynched case... at most, it's just a less serious problem. WFL is built as a pure-functional language; that's why there should be no side-effects. Even FAI functions that _appear_ to have side-effects actually just return an object representing the action to be performed, which is then executed _after_ the formula is fully evaluated. However, formula filters have no execution stage.
